g2p/transducer/__init__.py

```diff
@@ -1249,3 +1254,53 @@ def check(self, tg: TransductionGraph, shallow=False, display_warnings=False):
                 else:
                     return False
         return result
+
+
+def preserve_case(
+    tg: TransductionGraph, case_equivalencies: Dict[str, str] = None
+) -> TransductionGraph:
+    if case_equivalencies is None:
+        case_equivalencies = {}
+    reverse_case_equivalencies = {v: k for k, v in case_equivalencies.items()}
+    all_lower_case_equivalencies = case_equivalencies.keys()
+    all_upper_case_equivalencies = case_equivalencies.values()
+    new_string = ""
+    for item in tg.substring_alignments():
+        in_sub = item[0]
+        out_sub = item[1]
+        any_in_upper = any(x.isupper() for x in in_sub)
+        any_in_lower = any(x.islower() for x in in_sub)
+        any_out_upper = any(x.isupper() for x in out_sub)
+        any_out_lower = any(x.islower() for x in out_sub)
+        # continue if character is un-caseable
+        if (
+            out_sub not in case_equivalencies
+            and not any_out_upper
+            and not any_out_lower
+        ):
+            new_string += out_sub
+            continue
+        # lower case using case equivalencies if they exist
+        if (
+            any_in_lower or in_sub in all_lower_case_equivalencies
+        ) and out_sub in all_upper_case_equivalencies:
+            new_string += reverse_case_equivalencies[out_sub]
+            continue
+        # upper case using case equivalencies if they exist
+        elif (
+            any_in_upper or in_sub in all_upper_case_equivalencies
+        ) and out_sub in all_lower_case_equivalencies:
+            new_string += case_equivalencies[out_sub]
+            continue
+        # change to upper if required
+        if any_in_upper and any_out_lower:
+            new_string += out_sub.upper()
+            continue
+        # change to lower if required
+        if any_in_lower and any_out_upper:
+            new_string += out_sub.lower()
+            continue
+        # just in case, append the out_sub
+        new_string += out_sub
+    tg.output_string = new_string
+    return tg
```

Comment:
Boy, it's hard to follow all this logic! We need more extensive unit testing just to make sure it works.

E.g., exercise these:
 - Given rule `aa -> bb`, `Aaxyz` should get mapped to `Bbxyz`
 - the hard stuff in kwk BOAS with those small caps being mapped to a context-dependent case
 - Given rule `a -> bb` `AXYZ` should get mapped to `BBXYZ` but `Axyz` to `Bbxyz` (although this is actually not fully reliable, e.g., in Spanish in some cases `ll` is capitalized as a unit, e.g. `llama` might capitalize to `LLama` though from my experience that's not a uniformly applied rule. And does kwk Umista consider `tł` as a single letter capitalized jointly, or as two letters with only the T being uppercased in capitalized words?)

Reply:
> Boy, it's hard to follow all this logic! We need more extensive unit testing just to make sure it works.
> 
> E.g., exercise these:
> 
> * Given rule `aa -> bb`, `Aaxyz` should get mapped to `Bbxyz`

Yes, this works.

> * the hard stuff in kwk BOAS with those small caps being mapped to a context-dependent case

I don't fundamentally really understand what is going on here, and I can't reproduce the examples you have above in the rules even without changing anything so we need to talk more about this.

> * Given rule `a -> bb` `AXYZ` should get mapped to `BBXYZ` but `Axyz` to `Bbxyz` (although this is actually not fully reliable, e.g., in Spanish in some cases `ll` is capitalized as a unit, e.g. `llama` might capitalize to `LLama` though from my experience that's not a uniformly applied rule. And does kwk Umista consider `tł` as a single letter capitalized jointly, or as two letters with only the T being uppercased in capitalized words?)

I disagree that for a rule a -> bb Axyz should get turned to Bbxyz. We will handle this as BBxyz, otherwise someone would need to add a case equivalency for bb and Bb, in which case it should work.

Reply:
I guess we might need to ask Daisy what she thinks is right for kwk. And I guess we should also test with kwk and see what comes out with your current implementation.

So for kwk my assumption has been that you can infer whether a small cap L or a small cap E should be interpreted as upper case by the case of the next letter. I suppose we should validate that assumption with Daisy before investing too much development effort making it work...